File: app/src/main/java/com/locallinkonline/stcatherineschool/rest/tasks/GetNewDataResourcesTask.java
package com.locallinkonline.stcatherineschool.rest.tasks;

import android.app.job.JobParameters;
import android.app.job.JobService;

import com.locallinkonline.stcatherineschool.rest.model.DataResourceResponse;
import com.locallinkonline.stcatherineschool.room.dao.DataResourcesDao;
import com.locallinkonline.stcatherineschool.room.db.StCatherineDatabase;
import com.locallinkonline.stcatherineschool.room.entity.MenuEntity;

import java.io.IOException;
import java.util.Collections;
import java.util.List;

import retrofit2.Call;
import retrofit2.Response;

public class GetNewDataResourcesTask extends StCatherineTask<JobParameters> {

    private DataResourcesDao dataResourcesDao;

    private final JobService job;

    public GetNewDataResourcesTask(StCatherineDatabase db, String url, JobService job) {
        super(db, url);
        this.dataResourcesDao = db.dataResourcesDao();
        this.job = job;
    }

    @Override
    protected Void doInBackground(JobParameters... args) {
        Call<DataResourceResponse> dataResourceResponseCall = stCatherineApi.getDataResources();

        boolean success;

        try
        {
            Response<DataResourceResponse> response = dataResourceResponseCall.execute();
            List<MenuEntity> menuEntities = response.body() != null ? response.body().getResourceList() : Collections.EMPTY_LIST;

            if(menuEntities != null && menuEntities.size() > 0) {
                dataResourcesDao.updateMenuItems(menuEntities);
            }

            success = true;
        } catch (IOException e) {
            success = false;
        }

        if(args != null && args.length > 0) {
            job.jobFinished(args[0], !success);
        }

        return null;
    }
}

File: app/src/main/java/com/locallinkonline/stcatherineschool/rest/tasks/GetNewDataResourcesTask.kt
package com.locallinkonline.stcatherineschool.rest.tasks

import android.app.job.JobParameters
import android.app.job.JobService

import com.locallinkonline.stcatherineschool.rest.model.DataResourceResponse
import com.locallinkonline.stcatherineschool.room.dao.DataResourcesDao
import com.locallinkonline.stcatherineschool.room.db.StCatherineDatabase
import com.locallinkonline.stcatherineschool.room.entity.MenuEntity

import java.io.IOException
import java.util.Collections

import retrofit2.Call
import retrofit2.Response

class GetNewDataResourcesTask(db: StCatherineDatabase, url: String, private val job: JobService?) : StCatherineTask<JobParameters>(db, url) {

    private val dataResourcesDao: DataResourcesDao = db.dataResourcesDao()

    override fun doInBackground(vararg args: JobParameters): Void? {
        val dataResourceResponseCall = stCatherineApi.dataResources

        var success: Boolean

        try {
            val response = dataResourceResponseCall.execute()
            val menuEntities = response.body()!!.resourceList

            if (menuEntities != null && menuEntities.isNotEmpty()) {
                dataResourcesDao.updateMenuItems(menuEntities)
            }

            success = true
        } catch (e: IOException) {
            success = false
        }

        if (args.size > 0 &&job != null) {
            job.jobFinished(args[0], !success)
        }

        return null
    }
}
